### PR TITLE
golos.worker implementation

### PR DESCRIPTION
**Pull request by denis4net**

### CODE REVIEW

File: contracts/golos.worker/golos.worker.test.js [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+      code: "golos.worker",
+      scope: appName,
+      table: "proposals",
+      lower_bound: proposalId,
+      limit: 1
+    })).rows[0];
+}
+
+it(
+  "1st use case test",
```
Comment (zxcat) [inline contracts/golos.worker/golos.worker.test.js]: This test checks only "positive" scenario, when all actions and parameters are right. This is not enough to be sure contract works as we expect and throws where we expect it to throw.

Please add separate (unit) tests for each action, and check following:
1. throw of each assert of an action (+check message to ensure that expected assert throws).
2. precise values check: if assert checks `< x`, than we must check boundary: `x-1` succeed, `x` fails.
3. same on timing. if something changed at arbitrary time point, we should check it *one block before* that point, then `produce_blocks(1)` and check, that state changed as expected and when expected.
4. check auths: if action pushed by wrong account, it must fail.
5. check order: some actions must fail if called in wrong state.

File: contracts/golos.worker/main.cpp [hunk 2]
```diff
@@ -0,0 +1,1088 @@
...
+    {
+      data.update(that);
+      modified = TIMESTAMP_NOW;
+    }
+  };
+
+  typedef uint64_t proposal_id_t;
+
+  //@abi table proposals i64
+  struct proposal_t
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: this mega-struct is bad to have in contract. it can contain tons of comments and votes, so consume lot of RAM. and RAM payer changes on each action that changes data. so popular proposal auto-restricts accounts with low RAM to comment it. furthermore, serialize/deserialize of such struct is CPU-consuming, so at some point it can not fit in CPU bandwidth. one more point: searches in vectors of of inner structs do not use indexes and have O(N) complexity + lot of unnecessary reads.

Please, divide this struct to separate tables. Object creator (comment author, vote caster) should create it in his RAM (pay only for his objects)

Comment (denis4net) [reply contracts/golos.worker/main.cpp]: Is it critical for PoC?

File: contracts/golos.worker/golos.worker.test.js (hunk 1 above)
Comment (denis4net) [reply contracts/golos.worker/golos.worker.test.js]: Point me to the unit test example in your project.

File: contracts/golos.worker/main.cpp [hunk 3]
```diff
@@ -0,0 +1,1088 @@
...
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+
+    const auto tspec_ptr = get_tspec(*proposal_ptr, tspec_app_id);
+    eosio_assert(tspec_ptr != proposal_ptr->tspec_apps.end(), "technical specification applicatoin doesn't exist");
+    require_app_delegate(author);
+    eosio_assert(voting_time_s + tspec_ptr->created.to_time_point().sec_since_epoch() >= now(), "voting time is over");
+
+    get_proposals().modify(proposal_ptr, author, [&](auto &o) {
+      auto tspec = get_tspec(o, tspec_app_id);
+      tspec->votes.vote(author, static_cast<voting_module_t::vote_value_t>(vote));
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: this will not work as needed — votes are not approvals

File: contracts/golos.worker/main.cpp [hunk 4]
```diff
@@ -0,0 +1,1088 @@
...
+  void publishtspec(proposal_id_t proposal_id, const tspec_data_t &data)
+  {
+    LOG("proposal_id: %", proposal_id);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_CREATE, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->tspec_author);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->tspec_author, [&](proposal_t &proposal) {
+      proposal.tspec.update(data);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: don't allow pushing action with empty data (which changes nothing)

File: contracts/golos.worker/main.cpp [hunk 5]
```diff
@@ -0,0 +1,1088 @@
...
+   * @brief publishtspec publish a final tehcnical specification
+   * @param proposal_id proposal ID
+   * @param data technical specification details
+   */
+  /// @abi action
+  void publishtspec(proposal_id_t proposal_id, const tspec_data_t &data)
+  {
+    LOG("proposal_id: %", proposal_id);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_CREATE, "invalid proposal state");
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: *TODO: now it allows push this action several times. can be used to edit. is it needed?*

File: contracts/golos.worker/main.cpp [hunk 6]
```diff
@@ -0,0 +1,1088 @@
...
+    {
+      require_auth(proposal_ptr->worker);
+    }
+    else
+    {
+      require_auth(proposal_ptr->tspec_author);
+    }
+
+    get_proposals().modify(proposal_ptr, initiator, [&](proposal_t &proposal) {
+      refund(proposal, initiator);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: change proposal state to `cancelled`

File: contracts/golos.worker/main.cpp [hunk 7]
```diff
@@ -0,0 +1,1088 @@
...
+   * @param finished true if all work done
+   */
+  /// @abi action
+  void poststatus(proposal_id_t proposal_id, comment_id_t comment_id, const comment_data_t &comment, bool finished)
+  {
+    LOG("proposal_id: %, comment: %, final: %", proposal_id, comment.text.c_str(), (int) finished);
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_WORK, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->worker);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: tspec author can post status too

File: contracts/golos.worker/main.cpp [hunk 8]
```diff
@@ -0,0 +1,1088 @@
...
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->worker, [&](auto &proposal) {
+      proposal.work_status.add(comment_id, proposal.worker, comment);
+
+      if (finished)
+      {
+        proposal.set_state(proposal_t::STATE_TSPEC_AUTHOR_REVIEW);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: this stage is not needed, it's off-chain. worker and tspec author communicate, so here is no reason to worker to mark work finished. tspec author knows when work is finished and initiates "delegates review" from `STATE_WORK` state

File: contracts/golos.worker/main.cpp [hunk 9]
```diff
@@ -0,0 +1,1088 @@
...
+  {
+    LOG("proposal_id: %, comment: %", proposal_id, comment.text.c_str());
+    auto proposal_ptr = get_proposal(proposal_id);
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_AUTHOR_REVIEW, "invalid proposal state");
+    eosio_assert(proposal_ptr->type == proposal_t::TYPE_1, "unsupported action");
+    require_auth(proposal_ptr->tspec_author);
+
+    get_proposals().modify(proposal_ptr, proposal_ptr->tspec_author, [&](auto &proposal) {
+      proposal.set_state(proposal_t::STATE_DELEGATES_REVIEW);
+      proposal.work_status.add(comment_id, proposal.tspec_author, comment);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: this "comment" is different, not sure it fits in `work_status`. at least it must be marked somehow

File: contracts/golos.worker/main.cpp [hunk 10]
```diff
@@ -0,0 +1,1088 @@
...
+    get_proposals().modify(proposal_ptr, reviewer, [&](proposal_t &proposal) {
+      switch (status)
+      {
+      case proposal_t::STATUS_REJECT:
+        eosio_assert(proposal.state == proposal_t::STATE_DELEGATES_REVIEW ||
+                         proposal.state == proposal_t::STATE_WORK ||
+                         proposal.state == proposal_t::STATE_TSPEC_AUTHOR_REVIEW,
+                     "invalid state " __FILE__ ":" TOSTRING(__LINE__));
+
+        proposal.review_votes.downvote(reviewer);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: `downvote` method doesn't allow to change vote if upvote vas previously casted

File: contracts/golos.worker/main.cpp [hunk 11]
```diff
@@ -0,0 +1,1088 @@
...
+          //TODO: check that all voters are delegates in this moment
+          LOG("work has been rejected by the delegates voting, got % negative votes", proposal.review_votes.downvotes.size());
+          refund(proposal, reviewer);
+          close(proposal);
+        }
+        break;
+
+      case proposal_t::STATUS_ACCEPT:
+        eosio_assert(proposal_ptr->state == proposal_t::STATE_DELEGATES_REVIEW, "invalid state " __FILE__ ":" TOSTRING(__LINE__));
+        proposal.review_votes.upvote(reviewer);
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: `upvote` method doesn't allow to change vote if downvote vas previously casted

File: contracts/golos.worker/main.cpp [hunk 12]
```diff
@@ -0,0 +1,1088 @@
...
+    asset quantity;
+
+    if (proposal_ptr->tspec.payments_count == 1)
+    {
+      quantity = proposal_ptr->tspec.development_cost;
+    }
+    else
+    {
+      const uint32_t payment_interval = proposal_ptr->tspec.development_eta.to_time_point().sec_since_epoch() / proposal_ptr->tspec.payments_count;
+      const uint32_t payment_epoch = (now() - proposal_ptr->work_begining_time.to_time_point().sec_since_epoch()) / payment_interval;
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: use separate duration or interval field

File: contracts/golos.worker/golos.worker.test.js (hunk 1 above)
Comment (zxcat) [reply contracts/golos.worker/golos.worker.test.js]: https://github.com/GolosChain/golos-smart/blob/master/tests/golos.ctrl_tests.cpp

File: contracts/golos.worker/main.cpp [hunk 13]
```diff
@@ -0,0 +1,1088 @@
...
+    action(
+        permission_level{_self, N(active)},
+        TOKEN_ACCOUNT, N(transfer),
+        std::make_tuple(_self, proposal_ptr->worker,
+                        quantity, std::string("worker reward")))
+        .send();
+  }
+
+  // https://tbfleming.github.io/cib/eos.html#gist=d230f3ab2998e8858d3e51af7e4d9aeb
+  static void transfer(uint64_t code, currency::transfer &t)
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: Can you point me to the example, how esoio.token headers can be correctly included (compatible with eosio.cdt and eosiocpp from eosio docker container)?

File: contracts/golos.worker/main.cpp [hunk 14]
```diff
@@ -0,0 +1,1088 @@
...
+    print_f("%: transfer % from \"%\" to \"%\"", __FUNCTION__, t.quantity, ACCOUNT_NAME_CSTR(t.from), ACCOUNT_NAME_CSTR(t.to));
+
+    worker self(current_receiver(), eosio::string_to_name(t.memo.c_str()));
+    if (t.to != self._self || t.quantity.symbol == self.get_state().token_symbol || code != TOKEN_ACCOUNT)
+    {
+      return;
+    }
+
+    eosio_assert(t.quantity.is_valid(), "Quntity is invalid");
+    eosio_assert(t.quantity.amount > 0, "Invalid transfer amount");
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: fixed

File: contracts/golos.worker/main.cpp [hunk 15]
```diff
@@ -0,0 +1,1088 @@
...
+
+    const account_name &payer = t.to;
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: fixed

File: contracts/golos.worker/main.cpp [hunk 16]
```diff
@@ -0,0 +1,1088 @@
...
+namespace golos
+{
+using std::string;
+
+class worker : public contract
+{
+public:
+  static constexpr uint32_t voting_time_s = 7 * 24 * 3600;
+
+  typedef symbol_name app_domain_t;
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: fixed

File: contracts/golos.worker/main.cpp [hunk 17]
```diff
@@ -0,0 +1,1088 @@
...
+    {
+      self.get_funds().emplace(payer, [&](auto &fund) {
+        fund.owner = t.from;
+        fund.quantity = t.quantity;
+      });
+    }
+    else
+    {
+      self.get_funds().modify(fund, payer, [&](auto &fund) {
+        eosio_assert(fund.owner == t.from, "invalid fund owner");
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: It will not happen, you are right. I removed.

File: contracts/golos.worker/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (zxcat) [reply contracts/golos.worker/main.cpp]: This way is ok (for inline functions), but current `eosio.token` (and other contracts) do not declare structures of actions' arguments. Looks like the only way is to define `currency::transfer` replacement yourself. Not sure though

File: contracts/golos.worker/main.cpp (hunk 8 above)
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: A flag `finished` was discussed in the comment in the software requirements specification in google docs. Ok, I removed it.

File: contracts/golos.worker/main.cpp (hunk 9 above)
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: It was initially designed to be the last comment.

File: contracts/golos.worker/main.cpp (hunk 13 above)
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: I have got `transfer_args` from there https://github.com/EOSIO/eosio.contracts/blob/6ca72e709faba179726a20571929a9eeaea47d08/eosio.token/include/eosio.token/eosio.token.hpp#L66

File: contracts/golos.worker/golos.worker.test.js (hunk 1 above)
Comment (denis4net) [reply contracts/golos.worker/golos.worker.test.js]: > This test checks only "positive" scenario, when all actions and parameters are right. This is not enough to be sure contract works as we expect and throws where we expect it to throw.

Should I add a functional test for the negative scripts?

File: contracts/golos.worker/main.cpp (hunk 11 above)
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: Fixed

File: contracts/golos.worker/main.cpp (hunk 10 above)
Comment (denis4net) [reply contracts/golos.worker/main.cpp]: Fixed

File: contracts/golos.worker/golos.worker.test.js (hunk 1 above)
Comment (zxcat) [reply contracts/golos.worker/golos.worker.test.js]: > Should I add a functional test for the negative scripts?

It's not necessary, unit-tests will cover it

File: contracts/golos.worker/main.cpp (hunk 2 above)
Comment (zxcat) [reply contracts/golos.worker/main.cpp]: Yes

File: contracts/golos.worker/main.cpp [hunk 18]
```diff
@@ -934,7 +927,7 @@ class worker : public contract
   {
     LOG("proposal_id: %, comment: %", proposal_id, comment.text.c_str());
     auto proposal_ptr = get_proposal(proposal_id);
-    eosio_assert(proposal_ptr->state == proposal_t::STATE_TSPEC_AUTHOR_REVIEW, "invalid proposal state");
+    eosio_assert(proposal_ptr->state == proposal_t::STATE_DELEGATES_REVIEW, "invalid proposal state");
```
Comment (zxcat) [inline contracts/golos.worker/main.cpp]: looks like it's now impossible to change state to `STATE_DELEGATES_REVIEW`, here should be `STATE_WORK`

File: contracts/golos.worker/main.cpp (hunk 8 above)
Comment (zxcat) [reply contracts/golos.worker/main.cpp]: https://github.com/GolosChain/golos.worker/pull/1/commits/1a161585c2f0e3b0a033b47f899e09e0bb57b6b7 fixes this partly, please, update js too (wrong state enum, etc) and fix https://github.com/GolosChain/golos.worker/pull/1#discussion_r231797352

File: contracts/golos.worker/golos.worker_tests.cpp [hunk 19]
```diff
@@ -0,0 +1,555 @@
...
+    CHECK_EQUAL_OBJECTS(get_witness(_w[0]), wp("total_weight",(800+700+100+100)*1e6)("name","witn1"));
+    CHECK_MATCHING_OBJECT(get_witness(_w[1]), wp("total_weight",(800+100)*1e6)("name","witn2"));
+    produce_block();
+
+    // TODO: check decreasing vesting and paths, other than `issue`+`transfer`
+
+} FC_LOG_AND_RETHROW()
+
+
+BOOST_AUTO_TEST_SUITE_END()
```
Comment (zxcat) [inline contracts/golos.worker/golos.worker_tests.cpp]: looks like this file is not used

File: contracts/golos.worker/structs.hpp [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+ friend DataStream& operator >> ( DataStream& ds, TYPE& t ){ \
+    return ds >> static_cast<BASE&>(t); \
+ }
+
+namespace golos {
+
+using std::vector;
+
+template <typename T>
+class set_t : public vector<T>
```
Comment (zxcat) [inline contracts/golos.worker/structs.hpp]: is it used somewhere?

File: contracts/golos.worker/golos.worker.cpp [hunk 21]
```diff
@@ -0,0 +1,1036 @@
...
+        uint8_t state;
+
+        EOSLIB_SERIALIZE(proposal_t, (id)(author)(type)(title)(description)(fund_name)(deposit)(tspec_author)(tspec)(worker)(work_begining_time)(worker_payments_count)(created)(modified)(state));
+
+        uint64_t primary_key() const { return id; }
+        void set_state(state_t new_state) { state = new_state; }
+    };
+    multi_index<"proposals"_n, proposal_t> _proposals;
+
+    struct [[eosio::table("state")]] state_t
```
[the remark this answers is not in the document]
Comment (zxcat) [reply contracts/golos.worker/golos.worker.cpp]: @denis4net, ok, it makes sense

File: contracts/golos.worker/golos.worker.cpp [hunk 22]
```diff
@@ -0,0 +1,1036 @@
...
+        if (initiator == proposal_ptr->worker)
+        {
+            require_auth(proposal_ptr->worker);
+        }
+        else
+        {
+            require_auth(proposal_ptr->tspec_author);
+        }
+
+        _proposals.modify(proposal_ptr, initiator, [&](proposal_t &proposal) {
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/golos.worker.cpp]: I changed state to `closed`, is it ok?

File: contracts/golos.worker/structs.hpp (hunk 20 above)
Comment (denis4net) [reply contracts/golos.worker/structs.hpp]: deleted

File: contracts/golos.worker/golos.worker.cpp [hunk 23]
```diff
@@ -0,0 +1,1036 @@
...
+        eosio::name author;
+        uint8_t type;
+        string title;
+        string description;
+        eosio::name fund_name;
+        asset deposit;
+        ///< technical specification author
+        eosio::name tspec_author;
+        ///< technical specification data
+        tspec_data_t tspec;
```
[the remark this answers is not in the document]
Comment (denis4net) [reply contracts/golos.worker/golos.worker.cpp]: fixied

File: contracts/golos.worker/golos.worker_tests.cpp (hunk 19 above)
Comment (denis4net) [reply contracts/golos.worker/golos.worker_tests.cpp]: removed

Comment (denis4net) [reply contracts/golos.worker/golos.worker_tests.cpp]: why? 
```
Indicates the end of a test suite. Test suites can be nested. This macro should appear as many times as there is a BOOST_AUTO_TEST_SUITE.
```
